### ISSUE TITLE
Support for X-Sendfile

### ISSUE THREAD

**Issue by evert** [sabre-io/dav#120]
_Original author: ncas...@gmail.com (August 13, 2010 08:17:08)_

SabreDAV should support the X-Sendfile header. This would improve file delivery performance.
See:
- http://blog.lighttpd.net/articles/2006/07/02/x-sendfile (lighttpd)
- http://tn123.ath.cx/mod_xsendfile/ (Apache)

_Original issue: http://code.google.com/p/sabredav/issues/detail?id=69_
##

**Comment by PVince81** [comment]
Nice plugin.

You might want to also add "X-Sendfile2" and "X-Accel-Redirect" too, if that makes sense, like https://github.com/owncloud/core/blob/master/lib/private/files.php#L177

**Comment by PVince81** [comment]
The current implementation seems to assume that whoever is using the plugin must decide whether to enable it or not. Another approach would be to make the plugin auto-detect the situation based on either some given headers like "MOD_X_SENDFILE_ENABLED" or let the caller set flags through its API or constructor.

**Comment by dratini0** [comment]
Would you mind if I tried to implement something lile what is in owncloud? Maybe with a bit more flexibility.

**Comment by evert** [comment]
Not at all, but it has to fit the architectural design. If you have an idea about an approach, share! :) i want to release 2.2 soon so it would be nice to include it

**Comment by dratini0** [comment]
The basic idea would be to have two optional parameters on the plugin's constructor, ```$headerName``` and ```$filenamePrefix```. If these are present, the httpGet will just use the given header and prepend the prefix to the filename. If they are not given, httpGet will try to guess the right course of action by using the same environmental variables owncloud uses: ```MOD_X_SENDFILE_ENABLED```, ```MOD_X_SENDFILE2_ENABLED```, ```MOD_X_ACCEL_REDIRECT_ENABLED```, ```MOD_X_ACCEL_REDIRECT_PREFIX```. I think httpGet is the first time the request is available, but I will have to verify that. Of course I will be using ```Request::getRawServerValue```.

**Comment by dratini0** [comment]
Hmm, I will have to rethink this, since servers are so largely dirrerent. Apache is a walk in the park, but lighthttpd needs workarounds for range support, while nginx is probably better off with a prefixing than with anything else. Therefore what I propose would be having httpGet fire an event with the filename and the request to some autodetectors which call their own helpers with custom data. If the server user wants to disable autodetection, they can pass the server's name and an array to the constructor, and the httpGet will call the appropriate helper function, probably using events again. Since I am bad at explaining, I will implement it for apache pretty soon.

**Comment by divinity76** [comment]
> Note that @jcfischer's assumption that anything needs to be buffered in memory is wrong though, unless owncloud doesn't correctly use streams (which I believe it does).

actually it's PHP's fault, php's implementation of stream_copy_to_stream() works on a fread()/fwrite() loop and is kindof shitty implemented to boot [(ref1)](https://github.com/sabre-io/http/pull/119) [(ref2)](https://bugs.php.net/bug.php?id=77930), it got some improvements in PHP 7.4 (not yet released, NEWS file calls it `  . Fixed bug #77930 (stream_copy_to_stream should use mmap more often)`), but it's still not using [sendfile()](http://man7.org/linux/man-pages/man2/sendfile.2.html)

**Comment by staabm** [comment]
thx for your input.

I guess you should open a php-src bug (in case none is existant yet, for sendfile support).

regarding sabre-dav, I guess all is said in one of the above comments in this thread:

> 
> 
> I just created a new branch for this:
> 
> [fruux/sabre-dav@`sendfile`](https://github.com/fruux/sabre-dav/tree/sendfile) [](/fruux/sabre-dav/tree/HEAD@{2014-04-13T06:39:30Z})
> 
> Want to help? Help test!
> 
> A simple fileserver:
> 
>     1. `mkdir public`
> 
>     2. `chmod 777 public`
> 
>     3. create `server.php`, containing the following:
> 
> 
> ```
> <?php
> 
> namespace Sabre\DAV;
> 
> date_default_timezone_set('UTC');
> 
> $root = new FSExt\Collection(__DIR__ . '/public');
> $server = new Server($root);
> $server->addPlugin(
>    new XSendFile\Plugin()
> );
> $server->exec();
> 
> ?>
> ```
> 
>     1. Make sure you access this as something like `http://localhost/~mydir/sabredav/server.php/` and open it in a webdav client.
> 
>     2. Let me know if it worked ;)

**Comment by yevon** [comment]
Hi any news on this one? This would be really interesting. Not really clear if sabre can do anything about this, or maybe it's better to report it to the native php functions?
